### ISSUE TITLE
Open input blocks and never completed

### ISSUE THREAD

**Issue by Daniel63656** [atsushieno/ktmidi#92]
I try to receive messages from a connected usb MIDI device. 

```
val midiAccess = AndroidMidiAccess(this)
val midiInput = runBlocking {
                        midiAccess.openInput(midiAccess.inputs.first().id)
                    }
                    midiInput.setMessageReceivedListener(this@MainActivity)
```

However, the app becomes unresponsive whenever calling openInput. This happens on google pixel 7 (android-version 14).
I noticed this did not happen on my older Samsung galaxy S9 (same code)

**Comment by atsushieno** [comment]
Opening a device (either via ktmidi or Android MIDI API directly) involves Service connection that most likely involves main thread (it might depend on how Android MIDI service is implemented, but anything that involves Service usually does), and you are most likely blocking the main thread. If you try to invoke `openInput()` on a background thread it may work.

**Comment by Daniel63656** [comment]
If I do it in another thread then simply no connection is established until the cable is removed. I only switched to main thread to see what's going on.
This is likely caused by a permission problem as it works on an older phone. Newer android versions need the user to permit USB connection for each device that is plugged in think.

**Comment by atsushieno** [comment]
If you are trying to connect to a MIDI device that needs permission, you are supposed to provide enough permission (manage permission requests and approvals).

When we just deal with MIDI inputs, there is no required permission in MIDI API itself, like this app does not need any: https://github.com/atsushieno/resident-midi-keyboard/blob/main/app/src/main/AndroidManifest.xml

**Comment by Daniel63656** [comment]
This got changed in new android versions. You now need to grant permission
via pop up for each USB device to initiate data transfer.
Other apps like Midi connector ble also don't work on newer phones for that
reason.
I am trying to grant the permission and see if it works then

Atsushi Eno ***@***.***> schrieb am Do., 31. Okt. 2024, 08:09:

**Comment by atsushieno** [comment]
Yes, please let me know if it works or not. I would make sure that you are not trying to block the main thread while you are trying to (indirectly) show the permission approval dialog, as (IF that's still the case) the permission dialog also needs the main thread.

**Comment by atsushieno** [comment]
Without any code I can examine, I can at best suggest some tips to debug your app issue: run it on Android Studio in debugging mode, then interrupt (break) while you are waiting for the `openInput()`, and check all those running threads on the debugger. There will be some blocking operation that is waiting for something that is waiting for some action that you think that should be already dispatched.
MIDI connections are controlled by `MidiManager`, the client to the system internal MIDI Service, which usually requires main thread not blocked.

**Comment by JBramauer** [comment]
I'm running USB midi on a Pixel 8 with Android 15 and I'm not requesting any permissions - so I assume there's no need  as long as the device is class compliant and doesn't require a specific driver.

Your usage of the `CoroutineScope` looks suspicious, though. You should almost never need to create an instance yourself. In a Compose UI you can use `rememberLifecycleScope()`, in plain Android use your activity/fragment `lifecycleScope` and then push to the IO Dispatcher via `withContext(Dispatchers.IO)`.
